### PR TITLE
Add support for blkio read/write bps device

### PR DESCRIPTION
**Pull request by Mashimiao**
Signed-off-by: Ma Shimiao mashimiao.fnst@cn.fujitsu.com

### CODE REVIEW

**Comment by LK4D4** [comment]
@Mashimiao don't know why, but this continue to fail on jenkins. Could you try to rebase against master pls?

**Comment by Mashimiao** [comment]
@LK4D4, in daemon file I have imported `syscall`. But, why when making windows binary it always says 
`daemon\daemon.go:937: undefined: syscall.Stat_t
daemon\daemon.go:941: undefined: syscall.Stat
daemon\daemon.go:949: undefined: syscall.Stat`

**Comment by tiborvass** [comment]
For reference, same problem as in #14006 (comment) we need to have a policy on what to do with these cgroups flags.

File: contrib/completion/bash/docker [hunk 1]
```diff
@@ -1027,6 +1027,7 @@ _docker_run() {
 		--net
 		--pid
 		--publish -p
+		--read-bps-device
```
Comment (tiborvass) [inline contrib/completion/bash/docker]: @Mashimiao could this be `device-bps-read` for lexicographical order?

**Comment by tiborvass** [comment]
For the same reasons described in https://github.com/docker/docker/pull/14006#issuecomment-128468048 I think we should accept these flags.

**Comment by cpuguy83** [comment]
@tiborvass Should we go ahead and move to code-review?

**Comment by tiborvass** [comment]
@cpuguy83 was thinking we could wait for OP to answer to see if he disagrees.

File: contrib/completion/bash/docker (hunk 1 above)
Comment (Mashimiao) [inline contrib/completion/bash/docker]: Hi @tiborvass  --read-bps-device comes from cgroup file  blkio.throttle.read_bps_device. I think it's better to keep the same format with cgroup. If familiar with cgroup, it's easier for people to understand what they are going to set. But if you insisted, I think I'm OK to change it.

**Comment by cpuguy83** [comment]
ping @Mashimiao for Q regarding moving to code-review, also a rebase.

**Comment by Mashimiao** [comment]
ping @cpuguy83  rebased.

**Comment by jfrazelle** [comment]
sorry needs rebase

**Comment by Mashimiao** [comment]
ping @jfrazelle , @cpuguy83 , @icecrime  rabased and fixed.

File: daemon/container_unix.go [hunk 2]
```diff
@@ -223,6 +223,9 @@ func populateCommand(c *Container, env []string) error {
 		return err
 	}
 
+	readBpsDevice := strings.Join(c.hostConfig.BlkioReadBpsDevice, "\n")
```
Comment (cpuguy83) [inline daemon/container_unix.go]: How come these are being passed around as strings?

File: daemon/daemon_unix.go [hunk 3]
```diff
@@ -38,6 +38,38 @@ const (
 	platformSupported = true
 )
 
+func getBlkioReadBpsDevice(config *runconfig.HostConfig) ([]string, error) {
```
Comment (cpuguy83) [inline daemon/daemon_unix.go]: We should have a strong type for this instead of `[]string`

File: daemon/daemon_unix.go [hunk 4]
```diff
@@ -38,6 +38,38 @@ const (
...
+			return nil, err
+		}
+		ReadBpsDevice := fmt.Sprintf("%d:%d %s", uint64(stat.Rdev/256), uint64(stat.Rdev%256), splited[1])
+		BlkioReadBpsDevice = append(BlkioReadBpsDevice, ReadBpsDevice)
+	}
+
+	return BlkioReadBpsDevice, nil
+}
+
+func getBlkioWriteBpsDevice(config *runconfig.HostConfig) ([]string, error) {
```
Comment (cpuguy83) [inline daemon/daemon_unix.go]: We should have a strong type for this instead of `[]string`

File: daemon/execdriver/driver.go [hunk 5]
```diff
@@ -125,18 +125,20 @@ type UTS struct {
...
+	Memory                      int64            `json:"memory"`
+	MemorySwap                  int64            `json:"memory_swap"`
+	KernelMemory                int64            `json:"kernel_memory"`
+	CPUShares                   int64            `json:"cpu_shares"`
+	CpusetCpus                  string           `json:"cpuset_cpus"`
+	CpusetMems                  string           `json:"cpuset_mems"`
+	CPUPeriod                   int64            `json:"cpu_period"`
+	CPUQuota                    int64            `json:"cpu_quota"`
+	BlkioWeight                 int64            `json:"blkio_weight"`
+	BlkioThrottleReadBpsDevice  string           `json:"blkio_throttle_read_bps_device"`
```
Comment (cpuguy83) [inline daemon/execdriver/driver.go]: Should be a strong type with serialization to whatever format libcontainer expects (which would prefferably be a strong type as well, but I'm sure that's probably already in).

File: docs/reference/api/docker_remote_api_v1.20.md [hunk 6]
```diff
@@ -177,6 +177,8 @@ Create a container
              "CpusetCpus": "0,1",
              "CpusetMems": "0,1",
              "BlkioWeight": 300,
+             "BlkioReadBpsDevice": [""],
```
Comment (cpuguy83) [inline docs/reference/api/docker_remote_api_v1.20.md]: A strong type on these (like we do with ulimit) would be much nicer, IMO.

**Comment by tiborvass** [comment]
@Mashimiao sorry needs another rebase :(

**Comment by Mashimiao** [comment]
@tiborvass  rebased. and fixed according to #13959

**Comment by jfrazelle** [comment]
can we make the changes @cpuguy83 suggested

File: docs/reference/api/docker_remote_api_v1.20.md (hunk 6 above)
Comment (thaJeztah) [inline docs/reference/api/docker_remote_api_v1.20.md]: While changing this; these changes need to be moved to the API v1.21 docs

File: daemon/daemon_unix.go (hunk 3 above)
Comment (Mashimiao) [inline daemon/daemon_unix.go]: libcontainer accept this as string

File: daemon/container_unix.go (hunk 2 above)
Comment (Mashimiao) [inline daemon/container_unix.go]: libcontainer accept this as string

**Comment by runcom** [comment]
this is currently blocked by https://github.com/opencontainers/specs/pull/193 along with https://github.com/docker/docker/pull/15879 about read/write iops device

**Comment by tiborvass** [comment]
@runcom the PR you referenced is now merged

**Comment by runcom** [comment]
@tiborvass I'm waiting for the other runc PR I did https://github.com/opencontainers/runc/pull/276  to be merged which also includes the libcontainer fixes this PR needs. Once that is in I'll vendor libcontainer/runc here in docker and everything is good to go

File: docs/reference/run.md [hunk 7]
```diff
@@ -844,6 +846,16 @@ weights of the two containers.
 > **Note:** The blkio weight setting is only available for direct IO. Buffered IO
 > is not currently supported.
 
+The `--read-bps-device` flag can specify upper limit on READ rate from the device.
+For example, the command create one container and specify upper limit on READ rate from /dev/sda:
+
+    $ docker run -ti --name c1 --read-bps-device /dev/sda:300 ubuntu:14.04 /bin/bash
```
Comment (jfrazelle) [inline docs/reference/run.md]: this should use the "to human units" so for example `--read-bps-device /dev/sda:1mb` much like how the memory cgroups currently work

**Comment by jfrazelle** [comment]
can you add a test w a dd in a container and have `--read-bps-device /dev/zero:1mb`
so:
```
 dd if=/dev/zero of=/dev/null bs=4k count=1024             
# should be:
# 1024+0 records in
# 1024+0 records out
# 4194304 bytes (4.2 MB) copied, 4.0001 s, 1.0 MB/s
```

**Comment by jfrazelle** [comment]
@Mashimiao can you rebase and make the changes now that gh#16795 was merged

**Comment by Mashimiao** [comment]
Hi @jfrazelle, this PR is similar to #13959 and use some code from #13959. I want to rebase and fix this PR after #13959 is merged.

**Comment by jfrazelle** [comment]
No worries makes sense :)






On Thu, Oct 8, 2015 at 12:02 AM -0700, "Ma Shimiao" <notifications@github.com> wrote:












Hi @jfrazelle, this PR is similar to #13959 and use some code from #13959. I want to rebase and fix this PR after #13959 is merged.

**Comment by LK4D4** [comment]
@Mashimiao Looks like #13959 is merged now :)

**Comment by Mashimiao** [comment]
@LK4D4 rebased but janky failed with unrelated case.

**Comment by thaJeztah** [comment]
@Mashimiao I restarted Janky

File: docs/reference/run.md (hunk 7 above)
Comment (LK4D4) [inline docs/reference/run.md]: I agree

File: docs/reference/api/docker_remote_api_v1.20.md (hunk 6 above)
Comment (LK4D4) [inline docs/reference/api/docker_remote_api_v1.20.md]: I agree, it might be simpler to use datastructure for this.

**Comment by LK4D4** [comment]
@Mashimiao there are couple of comments from @cpuguy83 and @jfrazelle . Pls take a look.

File: daemon/daemon_unix.go (hunk 3 above)
Comment (Mashimiao) [inline daemon/daemon_unix.go]: fixed

File: daemon/daemon_unix.go (hunk 4 above)
Comment (Mashimiao) [inline daemon/daemon_unix.go]: fixed

File: daemon/execdriver/driver.go (hunk 5 above)
Comment (Mashimiao) [inline daemon/execdriver/driver.go]: fixed

File: docs/reference/api/docker_remote_api_v1.20.md (hunk 6 above)
Comment (Mashimiao) [inline docs/reference/api/docker_remote_api_v1.20.md]: Thanks, fixed.

**Comment by Mashimiao** [comment]
@LK4D4  all commetns fixed.

**Comment by LK4D4** [comment]
ping @cpuguy83

File: opts/throttledevice.go [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+		if err != nil {
+			return err
+		}
+		value = v
+	}
+	(opt.values) = append((opt.values), value)
+	return nil
+}
+
+// String returns Ulimit values as a string.
```
Comment (cpuguy83) [inline opts/throttledevice.go]: s/Ulimit/???/

File: contrib/completion/bash/docker (hunk 1 above)
Comment (cpuguy83) [inline contrib/completion/bash/docker]: I agree with @tiborvass here.
Most of the cgroup names are pretty sane, but this one is a bit backwards to me. I'd prefer to provide a better name, especially since this will be shared by other platforms.
This is just a nit...

We could probably even go a bit further and say `device-read-rate` and have it accept a user-friendly value  (e.g. `50mbps` or something instead of `52428800`).

File: opts/throttledevice.go (hunk 8 above)
Comment (Mashimiao) [inline opts/throttledevice.go]: fixed.

**Comment by runcom** [comment]
This needs another rebase :(

**Comment by Mashimiao** [comment]
rebased

File: daemon/daemon_windows.go [hunk 9]
```diff
@@ -24,6 +24,14 @@ func getBlkioWeightDevices(config *runconfig.HostConfig) ([]*blkiodev.WeightDevi
 	return nil, nil
 }
 
+func getBlkioReadBpsDevices(bpsDevices []string) ([]string, error) {
```
Comment (calavera) [inline daemon/daemon_windows.go]: I don't think we needs these definition in `_windows.go`. The caller of these functions is in a unix-only path. They are never called on windows, are they?

**Comment by calavera** [comment]
These two new fields need to be added to the API documentation, but that's for the next step. I left a minor comment about dead code that we can probably remove.

Otherwise, LGTM

**Comment by thaJeztah** [comment]
note: the documentation failures can probably be resolved by rebasing this PR

**Comment by Mashimiao** [comment]
@thaJeztah  rebased, but seems still failed

**Comment by thaJeztah** [comment]
@Mashimiao hm, not sure what the problem is with the documentation CI, it may be something that's still worked on (it's been recently added)

**Comment by Mashimiao** [comment]
@thaJeztah  rebased again. documentation to be success.

**Comment by jfrazelle** [comment]
LGTM

**Comment by jfrazelle** [comment]
ping @moxiegirl @thaJeztah for docs

File: docs/reference/commandline/run.md [hunk 10]
```diff
@@ -65,6 +65,7 @@ parent = "smn_cli"
       -p, --publish=[]              Publish a container's port(s) to the host
       --pid=""                      PID namespace to use
       --privileged=false            Give extended privileges to this container
+      --read-bps-device=""       Specifies upper limit on READ rate from the device
```
Comment (thaJeztah) [inline docs/reference/commandline/run.md]: looks like "Specifies" doesn't align with the other lines.

Perhaps we should use `Limit read rate from the device (in bits per second)` (or is it bytes per second?)

I wonder if we should be more specific about "device", it isn't clear to me *which* device; does it apply to any disk read (and write for `write-bps-device`)? Does it apply to volumes?

**Comment by thaJeztah** [comment]
Left one comment (but question applies to the wording in general)

We also need a note in the API changelog; https://github.com/docker/docker/blob/master/docs/reference/api/docker_remote_api.md#v122-api-changes

File: docs/reference/commandline/run.md (hunk 10 above)
Comment (thaJeztah) [inline docs/reference/commandline/run.md]: Oh, asked too soon; I see the example below. We probably need a "format" in the description, e.g.

```Markdown
The format is `<device-path>:<number>[unit]`. `number` must be greater than `0`. Unit is optional and can be `b` (bytes per second), `mb` (megabyte per second)....
```

Thinking out loud here, because I don't know if "unit" is optional, and which units are allowed :relaxed:

Comment (thaJeztah) [inline docs/reference/commandline/run.md]: Also, can this option be specified multiple times (for multiple devices?). If so, we should probably mention it as `--read-bps-device=[]` (square brackets)

File: docs/reference/api/docker_remote_api_v1.22.md [hunk 11]
```diff
@@ -244,6 +246,8 @@ Json Parameters:
 -   **CpusetMems** - Memory nodes (MEMs) in which to allow execution (0-3, 0,1). Only effective on NUMA systems.
 -   **BlkioWeight** - Block IO weight (relative weight) accepts a weight value between 10 and 1000.
  -   **BlkioWeightDevice** - Block IO weight (relative device weight) in the form of:        `"BlkioWeightDevice": [{"Path": "device_path", "Weight": weight}]`
+-   **BlkioReadBpsDevice** - Specifies upper limit on READ rate from the device.
+-   **BlkioWriteBpsDevice** - Specifies upper limit on WRITE rate to the device.
```
Comment (thaJeztah) [inline docs/reference/api/docker_remote_api_v1.22.md]: What's the format for this value? e.g. `[{"/dev/sda":"1234mb"}]`?

File: contrib/completion/zsh/_docker [hunk 12]
```diff
@@ -486,6 +486,7 @@ __docker_subcommand() {
         "($help)*"{-p=,--publish=}"[Expose a container's port to the host]:port:_ports"
         "($help)--pid=[PID namespace to use]:PID: "
         "($help)--privileged[Give extended privileges to this container]"
+        "($help)--read-bps-device=-[Specifies upper limit on READ rate from the device]:device:IO rate: "
```
Comment (moxiegirl) [inline contrib/completion/zsh/_docker]: Limits the read rate from a device <device-path>:<integer-limit>[unit]

File: contrib/completion/zsh/_docker [hunk 13]
```diff
@@ -494,6 +495,7 @@ __docker_subcommand() {
         "($help)*-v[Bind mount a volume]:volume: "
         "($help)--volume-driver=[Optional volume driver for the container]:volume driver:(local)"
         "($help)*--volumes-from=[Mount volumes from the specified container]:volume: "
+        "($help)--write-bps-device=-[Specifies upper limit on WRITE rate to the device]:device:IO rate: "
```
Comment (moxiegirl) [inline contrib/completion/zsh/_docker]: Limits the write rate to a device <device-path>:<integer-limit>[unit]

File: daemon/daemon_unix.go [hunk 14]
```diff
@@ -241,6 +271,16 @@ func verifyPlatformContainerSettings(daemon *Daemon, hostConfig *runconfig.HostC
 		logrus.Warnf("Your kernel does not support Block I/O weight_device. Weight-device discarded.")
 		hostConfig.BlkioWeightDevice = []*pblkiodev.WeightDevice{}
 	}
+	if len(hostConfig.BlkioReadBpsDevice) > 0 && !sysInfo.BlkioReadBpsDevice {
+		warnings = append(warnings, "Your kernel does not support Block read limit in bytes per second.")
```
Comment (moxiegirl) [inline daemon/daemon_unix.go]: This error message isn't clear --- is the flag not supported or simply the configuration the user provided? If the flag is not supported period because of the type of device. Hard to tell from the language. Some better choices that give users some better choices.

```
Your kernel does not support the --read-bps-device flag .
Your kernel does not support the --read-bps-device flag with the XXX device. 
You specified the --read-bps-device=xXXX:XXX values. Your kernel does not support this configuration. Specify another.
```

File: daemon/daemon_unix.go [hunk 15]
```diff
@@ -241,6 +271,16 @@ func verifyPlatformContainerSettings(daemon *Daemon, hostConfig *runconfig.HostC
...
+		logrus.Warnf("Your kernel does not support Block I/O read limit in bytes per second. Read-bps-device discarded.")
+		hostConfig.BlkioReadBpsDevice = []*pblkiodev.ThrottleDevice{}
+	}
+	if len(hostConfig.BlkioWriteBpsDevice) > 0 && !sysInfo.BlkioWriteBpsDevice {
+		warnings = append(warnings, "Your kernel does not support Block write limit in bytes per second.")
```
Comment (moxiegirl) [inline daemon/daemon_unix.go]: Similar problem here:

```
Your kernel does not support the --write-bps-device flag .
Your kernel does not support the --wite-bps-device flag with the XXX device. 
You specified the --wtie-bps-device=xXXX:XXX values. Your kernel does not support this configuration. Specify another.
```

File: daemon/daemon_unix.go [hunk 16]
```diff
@@ -241,6 +271,16 @@ func verifyPlatformContainerSettings(daemon *Daemon, hostConfig *runconfig.HostC
...
+		logrus.Warnf("Your kernel does not support Block I/O write limit in bytes per second. Write-bps-device discarded.")
```
Comment (moxiegirl) [inline daemon/daemon_unix.go]: If we are discarding the device, shouldn't this be an error? Why is one message discarding and the other not from the same `if` statement ....sorry my Go knowledge is lame perhaps this is normal.

File: docs/reference/run.md [hunk 17]
```diff
@@ -978,6 +980,16 @@ $ docker run -it \
     --blkio-weight-device "/dev/sda:200" \
     ubuntu
 
+The `--read-bps-device` flag can specify upper limit on READ rate from the device.
```
Comment (moxiegirl) [inline docs/reference/run.md]:
```markdown
The `--read-bps-device` flag limits the read rate from a device. For example,
this command creates a container and limits the read rate to `1mb` per XXXX
from `/dev/sda`:

    $ docker run -ti --name c1 --read-bps-device /dev/sda:1mb ubuntu:14.04 /bin/bash

The `--write-bps-device` flag limits the write rate to the device. For example,
this command creates a container and limits the write rate to `/dev/sda`: 

    $ docker run -ti --name c1 --write-bps-device /dev/sda:1mb ubuntu:14.04 /bin/bash

For both the read and write flag, you can specify the rate in XXX, XXX, or XXX.
```

File: docs/reference/commandline/run.md [hunk 18]
```diff
@@ -83,6 +84,7 @@ parent = "smn_cli"
                                     If neither 'rw' or 'ro' is specified then the volume is mounted
                                     in read-write mode.
       --volumes-from=[]             Mount volumes from the specified container(s)
+      --write-bps-device=""      Specifies upper limit on WRITE rate to the device
```
Comment (moxiegirl) [inline docs/reference/commandline/run.md]: Is there a reason for the all caps? Lowercase should be fine.

File: docs/reference/commandline/run.md (hunk 10 above)
Comment (moxiegirl) [inline docs/reference/commandline/run.md]: I think @thaJeztah is going in the right direction. 
```
| `--read-bps-device="<device-path>:<limit>[unit]" `    | Limit the read rate from a device.  Limit is a positive integer. Unit can be one of kb, mb, or gb.
```

File: docs/reference/commandline/create.md [hunk 19]
```diff
@@ -61,6 +61,7 @@ Creates a new container.
       -p, --publish=[]              Publish a container's port(s) to the host
       --pid=""                      PID namespace to use
       --privileged=false            Give extended privileges to this container
+      --read-bps-device=""          Specifies upper limit on READ rate from the device
```
Comment (moxiegirl) [inline docs/reference/commandline/create.md]:
```
| `--read-bps-device="<device-path>:<limit>[unit]" `    | Limit the read rate from a device.  Limit is a positive integer. Unit can be one of kb, mb, or gb.
| `--wite-bps-device="<device-path>:<limit>[unit]" `    | Limit the write rate to a device.  Limit is a positive integer. Unit can be one of kb, mb, or gb.
```

File: docs/reference/run.md [hunk 20]
```diff
@@ -624,6 +624,8 @@ container:
 | `--cpu-quota=0`            | Limit the CPU CFS (Completely Fair Scheduler) quota                                         |
 | `--blkio-weight=0`         | Block IO weight (relative weight) accepts a weight value between 10 and 1000.               |
 | `--blkio-weight-device=""` | Block IO weight (relative device weight, format: `DEVICE_NAME:WEIGHT`)                                                |
+| `--read-bps-device="" `    | Specifies upper limit on READ rate from the device.                                         |
+| --write-bps-device="" `    | Specifies upper limit on WRITE rate to the device.                                          |
```
Comment (moxiegirl) [inline docs/reference/run.md]: Missing tic here.  Also, is it necessary to actual specify that it is an _upper_ --- are their conditions under which I would specify the slowest a read or write rate allowed?  If no, "limit" itself should suffice:  

```
| `--read-bps-device="<device-path>:<limit>[unit]" `    | Limit the read rate from a device.  Limit is a positive integer. Unit can be one of kb, mb, or gb.
| `--write-bps-device="<device-path>:<limit>[unit]" `    | Limit the write rate to a device.  Limit is a positive integer. Unit can be one of kb, mb, or gb.
```

File: opts/opts.go [hunk 21]
```diff
@@ -210,6 +214,29 @@ func ValidateWeightDevice(val string) (*blkiodev.WeightDevice, error) {
...
+	}
+	if !strings.HasPrefix(split[0], "/dev/") {
+		return nil, fmt.Errorf("bad format for device path: %s", val)
+	}
+	rate, err := units.RAMInBytes(split[1])
+	if err != nil {
+		return nil, fmt.Errorf("invalid rate for device: %s", val)
+	}
+	if rate < 0 {
+		return nil, fmt.Errorf("invalid rate for device: %s", val)
```
Comment (moxiegirl) [inline opts/opts.go]:
```
Invalid rate for device: %s.  The correct format is <limit>:<unit>. Limit must be a positive integer. Unit is optional and can be kb, mb, or gb.
```

File: runconfig/parse.go [hunk 22]
```diff
@@ -110,6 +112,8 @@ func Parse(cmd *flag.FlagSet, args []string) (*Config, *HostConfig, *flag.FlagSe
 
 	cmd.Var(&flAttach, []string{"a", "-attach"}, "Attach to STDIN, STDOUT or STDERR")
 	cmd.Var(&flBlkioWeightDevice, []string{"-blkio-weight-device"}, "Block IO weight (relative device weight)")
+	cmd.Var(&flReadBpsDevice, []string{"-read-bps-device"}, "Specifies upper limit on READ rate from the device")
```
Comment (moxiegirl) [inline runconfig/parse.go]: []string{"--read-bps-device="<device-path>:<limit>[unit]" } , "Limits the read rate from a device.  Limit is a positive integer. Unit can be one of kb, mb, or gb."

[]string{"--write-bps-device="<device-path>:<limit>[unit]" } , "Limits the write rate to a device.  Limit is a positive integer. Unit can be one of kb, mb, or gb."

**Comment by moxiegirl** [comment]
@jfrazelle I don't see a test in here either...you mentioned having one earlier. You OK without it?

**Comment by jfrazelle** [comment]
would be niceee

**Comment by jfrazelle** [comment]
I actually really think we need one

**Comment by icecrime** [comment]
@jfrazelle Can somebody take it in a followup PR? This has been opened for quite some months already :-(

**Comment by jfrazelle** [comment]
Ya as long as we make an issue so we don't forget I can make a test

File: docs/reference/commandline/run.md (hunk 10 above)
Comment (Mashimiao) [inline docs/reference/commandline/run.md]: I think the words will be  too long if we want to describe  clearly arg fomat. How about just give an example like option --label

File: opts/opts.go (hunk 21 above)
Comment (Mashimiao) [inline opts/opts.go]: fixed

File: runconfig/parse.go (hunk 22 above)
Comment (Mashimiao) [inline runconfig/parse.go]: I used an example

File: docs/reference/run.md (hunk 20 above)
Comment (Mashimiao) [inline docs/reference/run.md]: fixed.

**Comment by Mashimiao** [comment]
rebased and fixed.

**Comment by runcom** [comment]
is this good to merge (given the lack of tests which will be addressed in another PR, I can take care of it also)?

File: contrib/completion/zsh/_docker [hunk 23]
```diff
@@ -462,6 +462,8 @@ __docker_subcommand() {
         "($help)*--cap-drop=[Drop Linux capabilities]:capability: "
         "($help)--cidfile=[Write the container ID to the file]:CID file:_files"
         "($help)*--device=[Add a host device to the container]:device:_files"
+        "($help)--device-read-bps=[Limit the read rate (bytes per second) from a device]:device:IO rate: "
+        "($help)--device-write-bps=[Limit the write rate (bytes per second) to a device]:device:IO rate: "
```
Comment (thaJeztah) [inline contrib/completion/zsh/_docker]: Just saw @sdurrheimer comment in your other PR (yay, I'm learning!) these should probably have a a `*` in front of `--`, so `"($help)*--device...`

File: docs/reference/api/docker_remote_api_v1.22.md [hunk 24]
```diff
@@ -245,6 +247,10 @@ Json Parameters:
 -   **CpusetMems** - Memory nodes (MEMs) in which to allow execution (0-3, 0,1). Only effective on NUMA systems.
 -   **BlkioWeight** - Block IO weight (relative weight) accepts a weight value between 10 and 1000.
 -   **BlkioWeightDevice** - Block IO weight (relative device weight) in the form of:        `"BlkioWeightDevice": [{"Path": "device_path", "Weight": weight}]`
+-   **BlkioDeviceReadBps** - Limit read rate from a device in form of:	`"BlkioDeviceReadBps": [{"Path": "device_path", "Rate": rate}]`, for example:
+	`"BlkioDeviceWriteBps": [{"Path": "/dev/sda", "Rate": "1024"}]"`
```
Comment (thaJeztah) [inline docs/reference/api/docker_remote_api_v1.22.md]: s/BlkioDeviceWriteBps/BlkioDeviceReadBps/

File: docs/reference/run.md [hunk 25]
```diff
@@ -978,6 +982,16 @@ $ docker run -it \
...
+The `--device-read-bps` flag can limit read rate from a device.
+For example, the command creates one container and limits theread rate to `1mb` per second from `/dev/sda`:
+
+    $ docker run -ti --name c1 --device-read-bps /dev/sda:1mb ubuntu:14.04 /bin/bash
+
+The `--device-write-bps` flag can limit write rate to a device.
+For example, the command creates one container and limits write rate to `1mb` per second to `/dev/sda`:
+
+    $ docker run -ti --name c1 --device-write-bps /dev/sda:1mb ubuntu:14.04 /bin/bash
+
```
Comment (thaJeztah) [inline docs/reference/run.md]: looks like you missed updating this with the changes @moxiegirl gave. Can you update this to
(I also slightly simplified the examples and removed irrelevant options);

```Markdown
The `--read-bps-device` flag limits the read rate from a device. For example,
this command creates a container and limits the read rate to `1mb` per second
from `/dev/sda`:

    $ docker run -it --read-bps-device /dev/sda:1mb ubuntu

The `--write-bps-device` flag limits the write rate to the device. For example,
this command creates a container and limits the write rate to `/dev/sda`: 

    $ docker run -it --write-bps-device /dev/sda:1mb ubuntu

Both flags take limits in the `<device-path>:<limit>[unit]` format. Both read
and write rates must be a positive integer. You can specify the rate in `kb`
(kilobytes), `mb` (megabytes), or `gb` (gigabytes).
```

**Comment by thaJeztah** [comment]
Thanks @Mashimiao, I left some comments, but we're almost there!

**Comment by thaJeztah** [comment]
Thanks @Mashimiao LGTM

@runcom will work on a test;  I saw some tiny issues with the docs, but I'll create a follow up PR to fix those. Thank you for contributing!

**Comment by moylop260** [comment]
Is https://github.com/docker/docker/issues/20077 a expected behaivor?

**Comment by wurstbrot** [comment]
Is it possible to use the parameter device-read-bps/device-read-bps for `docker run` in combination with docker-compose some how?
